test/fail_compilation/fail276.d

```diff
@@ -2,7 +2,7 @@
 TEST_OUTPUT:
 ---
 fail_compilation/fail276.d(19): Error: `this` has no effect
-fail_compilation/fail276.d(15): Error: `this` is only defined in non-static member functions, not `__anonclass2`
+fail_compilation/fail276.d(15): Error: cannot construct nested class `__anonclass1` because no implicit `this` reference to outer class `__anonclass2` is available
```

Comment:
Would it be better not to add names of anonymous classes?

Reply:
I'd vote for including the name of the anonymous class but adding a supplemental error to indicate that it is definitely generated by the language - the name will show up when debugging or in mangled names (etc. etc.) so best to opt for the most detail IMO.

Reply:
No, we should not leak compiler generated names to the user. I suggest to special case anonymous classes: "cannot construct anonymous class because no ...".

Reply:
ping @wolframw

Reply:
fixed